packages/xod-deploy/src/cloudCompiler.js

```diff
@@ -0,0 +1,132 @@
+import R from 'ramda';
+import path from 'path';
+import { outputFile } from 'fs-extra';
+import WebSocket from 'ws';
+import atob from 'atob';
+
+import { COMPILATION_RESPONSE_TYPES, COMPILATION_ERRORS } from './constants';
+
+// CompileError :: { errorCode: String, message: String }
+
+const createCompileError = (message, errorCode, payload = {}) => Object.assign(
+  new Error(message), { errorCode }, payload
+);
+
+/**
+ * Sends board identifier and code to compilation server
+ * and receives compiled binary file.
+ * Returns Promise with compiled binary file and filename or
+ * object with required fields: `type` and `message` and optional
+ * `code`, `error`, `timeout`.
+ */
+// :: { board: String, body: String } -> Promise { data: String, name: String } CompileError
+const sendAndReceive = dataToSend => new Promise((resolve, reject) => {
+  const ws = new WebSocket('wss://compile.xod.show/compile', {
+    perMessageDeflate: true,
+    handshakeTimeout: 3000,
+  });
+  ws.on('open', () => ws.send(JSON.stringify(dataToSend)));
+  ws.on('message', (res) => {
+    const data = JSON.parse(res);
+
+    const errCode = R.cond([
+      [
+        R.equals(COMPILATION_RESPONSE_TYPES.FAILED),
+        R.always(COMPILATION_ERRORS.COMPILE_FAILED),
+      ],
+      [
+        R.equals(COMPILATION_RESPONSE_TYPES.TIMEDOUT),
+        R.always(COMPILATION_ERRORS.COMPILE_TIMEDOUT),
+      ],
+      [
+        R.T,
+        R.always(null),
+      ],
+    ])(data.type);
+
+    if (errCode) {
+      reject(createCompileError(data.body, errCode));
+    }
+
+    if (data.type === COMPILATION_RESPONSE_TYPES.SUCCEEDED) {
```

Comment:
Could there be a situation when `errCode` is falsy and `type  !== COMPILATION_RESPONSE_TYPES.SUCCEEDED `?

Reply:
Yes. Added a comment about it.